### ISSUE TITLE
Reintegrate chalk into rustc

### ISSUE THREAD

**Issue by jackh726** [rust-lang/compiler-team#289]
# TL;DR

Reintegrate chalk into rustc using `chalk-solve` under the `-Z chalk` flag.

# Links and Details

This is working towards the eventual™ goal of using Chalk as the rustc trait solver. There used to be an experimental implementation of Chalk in rustc using `chalk-engine`. It became outdated and was removed in https://github.com/rust-lang/rust/pull/69247. Since that first experimental implementation was added, Chalk has gone through lots of changes, including adding a new crate, `chalk-solve`, which is supposed to implement Rust logic. As such, with this newer implementation, the interface between rustc and Chalk is a bit higher-level.

The PR for this initial MVP implementation is here: https://github.com/rust-lang/rust/pull/69406

# Mentors or Reviewers

c.c. @nikomatsakis
c.c. @rust-lang/wg-traits

**Comment by nikomatsakis** [comment]
@rustbot second

I'm going to exercise the "I consider this unlikely to be controversial clause" and r+ rust-lang/rust#69406 before it bitrots...we can always back it out. (Besides, if I know bors, it'll take 2 weeks to land anyway 😛)

**Comment by nikomatsakis** [comment]
Closing, FCP expired. =)
